File: 2306-create-binary-tree-from-descriptions/create-binary-tree-from-descriptions.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode createBinaryTree(int[][] descriptions) {
        Map<Integer, TreeNode> nodes = new HashMap<>();
        Set<Integer> children = new HashSet<>();

        for (int[] description : descriptions) {
            int parent = description[0];
            int child = description[1];
            int isLeft = description[2];

            nodes.putIfAbsent(parent, new TreeNode(parent));
            nodes.putIfAbsent(child, new TreeNode(child));

            if (isLeft == 1) {
                nodes.get(parent).left = nodes.get(child);
            } else {
                nodes.get(parent).right = nodes.get(child);
            }

            children.add(child);
        }

        for (int key : nodes.keySet()) {
            if (!children.contains(key)) {
                return nodes.get(key);
            }
        }

        return null;
    }
}

File: 2306-create-binary-tree-from-descriptions/create-binary-tree-from-descriptions.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def createBinaryTree(self, descriptions: List[List[int]]) -> Optional[TreeNode]:
        child = set()
        nodes = dict()

        for p, c, l in descriptions:
            if p not in nodes:
                nodes[p] = TreeNode(val=p)

            if c not in nodes:
                nodes[c] = TreeNode(val=c)

            if l == 1:
                nodes[p].left = nodes[c]
            else:
                nodes[p].right = nodes[c]

            child.add(c)

        for root in (set(nodes.keys()) - child):
            return nodes[root]
        
        return None
            